SIMILARITY RATE DETECTION WITH JACCARD AND COSINE

In [7]:
import nltk
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
nltk.download('stopwords')
from sklearn.metrics.pairwise import cosine_similarity



[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [8]:
def load_document(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        document = file.read()
    return document

def calculate_jaccard_similarity(doc1, doc2):
    words_doc1 = set(doc1.split())
    words_doc2 = set(doc2.split())
    intersection = words_doc1.intersection(words_doc2)
    union = words_doc1.union(words_doc2)
    similarity = len(intersection) / len(union)
    return similarity

def calculate_cosine_similarity(doc1, doc2):
    stop_words = set(stopwords.words('english'))
    vectorizer = TfidfVectorizer(stop_words=stop_words)
    tfidf_matrix = vectorizer.fit_transform([doc1, doc2])
    similarity = (tfidf_matrix * tfidf_matrix.T).A[0, 1]
    return similarity

# Step 1: Load the first document
doc1 = load_document('document1.txt')

# Step 2: Load the second document
doc2 = load_document('document2.txt')

# Step 3: Select the similarity algorithm
similarity_algorithm = input("Enter the similarity algorithm (Jaccard or Cosine): ")

# Step 4: Calculate the similarity and display the result
if similarity_algorithm.lower() == 'jaccard':
    similarity_score = calculate_jaccard_similarity(doc1, doc2)
elif similarity_algorithm.lower() == 'cosine':
    similarity_score = calculate_cosine_similarity(doc1, doc2)
else:
    print("Invalid similarity algorithm!")

print("Similarity score:", similarity_score)

Enter the similarity algorithm (Jaccard or Cosine): jaccard
Similarity score: 0.30177514792899407


In [10]:
def load_document(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        document = file.read()
    return document

def calculate_jaccard_similarity(doc1, doc2):
    words_doc1 = set(doc1.split())
    words_doc2 = set(doc2.split())
    intersection = words_doc1.intersection(words_doc2)
    union = words_doc1.union(words_doc2)
    similarity = len(intersection) / len(union)
    return similarity

def calculate_cosine_similarity(doc1, doc2):
    vectorizer = TfidfVectorizer(stop_words='english')
    tfidf_matrix = vectorizer.fit_transform([doc1, doc2])
    similarity_matrix = cosine_similarity(tfidf_matrix[0], tfidf_matrix[1])
    similarity_score = similarity_matrix[0][0]
    return similarity_score

# Step 1: Load the first document
doc1 = load_document('document1.txt')

# Step 2: Load the second document
doc2 = load_document('document2.txt')

# Step 3: Select the similarity algorithm
similarity_algorithm = input("Enter the similarity algorithm (Jaccard or Cosine): ")

# Step 4: Calculate the similarity and display the result
if similarity_algorithm.lower() == 'jaccard':
    similarity_score = calculate_jaccard_similarity(doc1, doc2)
elif similarity_algorithm.lower() == 'cosine':
    similarity_score = calculate_cosine_similarity(doc1, doc2)
else:
    print("Invalid similarity algorithm!")

print("Similarity score:", similarity_score)

Enter the similarity algorithm (Jaccard or Cosine): cosine
Similarity score: 0.6406191364251326


As a result We noticed that Documents', are related Dental Implants, Jaccard similarity rate------>0.30   
Cosine similarity rate------>0.64